### ISSUE TITLE
[BUG]Stopped uploading media data while many frames were lost.

### ISSUE THREAD

**Issue by August88618** [awslabs/amazon-kinesis-video-streams-producer-c#249]
When I stopped uploading media data at time A, the debug log showed that many frames were lost.
Q1: Is there any way to solve this problem? 
Q2: Is there any corresponding interface that can get the sending status of SDK, for example, whether it is still being sent?

[missingFrame_debuginfo.txt](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/files/7224875/missingFrame_debuginfo.txt)

**Comment by MushMal** [comment]
When you call stopSync, the SDK attempts to stream out the accumulated buffer then unblock. StopSync can be a lengthy operation and the timeout is controlled by the DeviceInfo.ClientInfo.stopTimeout param. If the API times out, it would return a 0x0000000f status. If the application then proceeds with freeing the stream, the stream will drop the context of the buffer and will log the information.

There is a dropped frame callback that will be called when frames are dropped with the information about the frame.

Your application can also call the PIC API to get the current window duration information - the upper layers call that API to understand the buffer size/duration.

**Comment by August88618** [comment]
if the stream is expected to stop after 10s, and at this point, there are still unsent frames in the cache, which are directly discarded, so the length of the video is not my expected video length, i don’t know when the sdk can be uploaded completely, so I can not control at what time to stop and free the stream. 
And according to your reply, one way is to use stoptimeout to control the stop execution timeout time to obtain the current stopstream timeout status, and the other way is to confirm whether the transmission is complete through the metric buffer size, as you said, this will be a time gap between the last  stream stop and the next start stream，the stop and the next start are within 1s in current demand scenario. If i want to ensure that the previous upload is complete, i must postpone the start time of the next recording. Is there any way to solve this problem for this scenario?

**Comment by MushMal** [comment]
I am really vague about your scenario but I am thinking you are attempting to upload video chunks. You could either control the stop stream timeout to ensure all the bits are sent, use the buffer size metric (not recommend in this scenario as you might have sent all but not yet received the ACKs) or another approach is to expect the ACK from the last fragment - in fact, this is what unblocks the sync version of the stop stream anyway.

Fundamentally, if you have an issue with the congestion, you need to resolve it at the application logic. For example, if you are still uploading the first chunk when your second chunk comes in, your application has to make decisions on what to do.

**Comment by August88618** [comment]
ths for your reply，the drop frame problem was solved using stopstreamSync instead of stopstream, so that the frame could keep putting after stopstream.
There is another question that When the SDK sends MKV cache data, the network is down, which interfaces should be called to
prolong its retransmission time or retransmission times to ensure complete upload of cached data?

**Comment by MushMal** [comment]
It really depends. For short-term outages you can control with the value of the stop stream timeout. When stop stream sync is called, the SDK is still attempting to stream out the entire buffer and receive the final fragment ACK.

For longer outages, you need to handle it in your application. For example, you could store the frames in a temp file and replay them in a new stream object after the network is back. SDK doesn't handle this level of application specific logic.

**Comment by disa6302** [comment]
Closing assuming question answered. @August88618 , please feel free to reopen if you have any more questions.
